### PR TITLE
US133005 - Convert d2l-rubric-loa-overlay to LitElement

### PR DESCRIPTION
**Pull request by semibran**
Start from [736429d](https://github.com/Brightspace/d2l-rubric/pull/1347/commits/736429d5c289c0456a2dd1a80f2f404975226e40) - can't seem to get the overall diff with the polymer version to cooperate.

I'm trying not to refactor too hard in this PR because actually trying to fix this component would involve getting way too in the weeds with rewriting abstractions around hypermedia.

Basically there's two main entities at play here, the LOA overlay entity and the LOA mapping entity. The LOA overlay entity is the parent entity - it gets the sequence of rubric levels and a link to the LOA mapping entity, which lists the actual LOA levels that are displayed in the overlay. Since the interactivity layer for the component involves both entity types there's an obscene number of href pojo maps and navigation methods baked into the overlay that'd need to be extracted out. So trying to refactor the component without breaking anything or just rebuilding from the ground up feels like it's turned into a bit of a house of cards situation :disappointed: It'd just take a surgical level of precision that I don't think belongs in a conversion ticket.

In any case I'm starting to think we need an idiom for wrapper interfaces around multiple hypermedia entities/entity collections. This sort of client-side entity interplay sheds light on a pretty glaring weakness in how we've been using HM in our webcomponents considering this isn't the only place where it happens in rubrics. I'll see if there's something I can come up with to address that in a separate PR

### CODE REVIEW

File: src/entities/rubric-level-entity.js [hunk 1]
```diff
@@ -46,14 +46,26 @@ export class RubricLevelEntity extends EntityBase {
 		return this._entity?.hasClass(Classes.rubrics.overridden) || false;
 	}
 
+	get loaHref() {
+		return this._entity?.getLinkByRel('https://achievements.api.brightspace.com/rels/level')?.href || '';
```
Comment (semibran) [inline src/entities/rubric-level-entity.js:50]: working on adding HM constants for this stuff

File: src/components/editor/rubric-criteria-editor.js [hunk 2]
```diff
@@ -359,7 +359,7 @@ class RubricCriteriaEditor extends composeMixins(
 					.token="${this.token}"
 					.context="${this.context}"
 					rubric-levels-href="${this._rubricLevelsHref}"
-					.rubric-level-loa-mapping="${this.rubricLevelLoaMapping}"
+					.rubric-level-loa-mapping="${JSON.stringify(this.rubricLevelLoaMapping) || {}}"
```
Comment (imcdonald-d2l) [inline src/components/editor/rubric-criteria-editor.js:362]: Why is this being converted to a string?

File: src/components/editor/rubric-loa-overlay.js [hunk 3]
```diff
@@ -0,0 +1,908 @@
...
+	EntityMixin(RubricLoaOverlayEntity),
+	EntityWatcherMixin,
+	LocalizeMixin,
+	TelemetryMixin,
+) {
+
+	static get properties() {
+		return {
+			hasOutOf: { type: Boolean, attribute: 'has-out-of' },
+			isHolistic: { type: Boolean, attribute: 'is-holistic' },
```
Comment (imcdonald-d2l) [inline src/components/editor/rubric-loa-overlay.js:30]: Looks like you're missing `reflect: true`. It is used in styles.

File: src/components/editor/rubric-loa-overlay.js [hunk 4]
```diff
@@ -0,0 +1,908 @@
...
+					width: 2.6rem;
+					border: none;
+					background-color: #ffffff;
+				}
+
+				.col-first[is-holistic] {
+					padding-left: 0;
+				}
+
+				:dir(rtl) .col-first[is-holistic] {
```
Comment (imcdonald-d2l) [inline src/components/editor/rubric-loa-overlay.js:99]: Should be `:host-context([dir='rtl'])` in a bunch of places. Will also need the `RtlMixin`

File: src/components/editor/rubric-loa-overlay.js [hunk 5]
```diff
@@ -0,0 +1,908 @@
...
+		}
+
+		this.requestUpdate('outcomeAlignments', oldOutcomeAlignments);
+	}
+
+	connectedCallback() {
+		super.connectedCallback();
+		document.addEventListener('mousemove', this._onMouseMove.bind(this));
+		document.addEventListener('mouseup', this._onMouseUp.bind(this));
+		document.addEventListener('d2l-loa-overlay-refresh', this._reloadEntity);
```
Comment (imcdonald-d2l) [inline src/components/editor/rubric-loa-overlay.js:243]: Do you not need to bind `this` to `_reloadEntity`? It is referencing `this.token` and `this.href`

File: src/components/editor/rubric-criteria-editor.js (hunk 2 above)
Comment (semibran) [reply src/components/editor/rubric-criteria-editor.js:362]: It's a lit gimmick I isolated recently around object propagation. `.property` binding works if the property name is one word like `context`, but if it's multiple words the child just won't receive the attribute. I wrote a blurb about it on confluence - basically the simplest way I've found to get it to work is to pass the property down as a string and receive it as an object. It deserializes for you without having to manually bake in any of the special converter stuff. [Lit playground](https://lit.dev/playground/#project=W3sibmFtZSI6InRlc3QtcGFyZW50LmpzIiwiY29udGVudCI6ImltcG9ydCB7aHRtbCwgTGl0RWxlbWVudH0gZnJvbSAnbGl0JztcbmltcG9ydCAnLi90ZXN0LWNoaWxkLmpzJztcblxuY3VzdG9tRWxlbWVudHMuZGVmaW5lKCd0ZXN0LXBhcmVudCcsIGNsYXNzIFRlc3RQYXJlbnQgZXh0ZW5kcyBMaXRFbGVtZW50IHtcbiAgc3RhdGljIHByb3BlcnRpZXMgPSB7XG4gICAgc3RyaW5nOiB7dHlwZTogU3RyaW5nfSxcbiAgICBvYmplY3Q6IHt0eXBlOiBPYmplY3R9LFxuICAgIHZlcmJvc2VTdHJpbmc6IHsgdHlwZTogU3RyaW5nIH0sXG4gICAgdmVyYm9zZU9iamVjdDogeyB0eXBlOiBPYmplY3QgfSxcbiAgICB2ZXJib3NlT2JqZWN0U2VyaWFsaXplZDogeyB0eXBlOiBPYmplY3QgfSxcbiAgfTtcblxuICBjb25zdHJ1Y3RvcigpIHtcbiAgICBzdXBlcigpO1xuICAgIHRoaXMuc3RyaW5nID0gJ2Zvbyc7XG4gICAgdGhpcy5vYmplY3QgPSB7IGhlbGxvOiAnd29ybGQnIH07XG4gICAgdGhpcy52ZXJib3NlU3RyaW5nID0gJ2hlbGxvIHdvcmxkISc7XG4gICAgdGhpcy52ZXJib3NlT2JqZWN0ID0ge1xuICAgICAgZm9vOiAzLFxuICAgICAgYmFyOiA0MixcbiAgICB9O1xuICAgIHRoaXMudmVyYm9zZU9iamVjdFNlcmlhbGl6ZWQgPSB7XG4gICAgICB0aGlzOiAnd29ya3MnLFxuICAgIH07XG4gIH1cblxuICByZW5kZXIoKSB7XG4gICAgcmV0dXJuIGh0bWxgXG4gICAgICA8dGVzdC1jaGlsZFxuICAgICAgICBzdHJpbmc9XCIke3RoaXMuc3RyaW5nfVwiXG4gICAgICAgIC5vYmplY3Q9XCIke3RoaXMub2JqZWN0fVwiXG4gICAgICAgIHZlcmJvc2Utc3RyaW5nPVwiJHt0aGlzLnZlcmJvc2VTdHJpbmd9XCJcbiAgICAgICAgLnZlcmJvc2Utb2JqZWN0PVwiJHtKU09OLnN0cmluZ2lmeSh0aGlzLnZlcmJvc2VPYmplY3QpfVwiXG4gICAgICAgIHZlcmJvc2Utb2JqZWN0LXNlcmlhbGl6ZWQ9XCIke0pTT04uc3RyaW5naWZ5KHRoaXMudmVyYm9zZU9iamVjdFNlcmlhbGl6ZWQpfVwiXG4gICAgICA8L3Rlc3QtY2hpbGQ-XG4gICAgYDtcbiAgfVxufSk7XG4ifSx7Im5hbWUiOiJpbmRleC5odG1sIiwiY29udGVudCI6IjwhRE9DVFlQRSBodG1sPlxuPGhlYWQ-XG4gIDxzY3JpcHQgdHlwZT1cIm1vZHVsZVwiIHNyYz1cIi4vdGVzdC1wYXJlbnQuanNcIj48L3NjcmlwdD5cbjwvaGVhZD5cbjxib2R5PlxuICA8dGVzdC1wYXJlbnQ-PC90ZXN0LXBhcmVudD5cbjwvYm9keT5cbiJ9LHsibmFtZSI6InBhY2thZ2UuanNvbiIsImNvbnRlbnQiOiJ7XG4gIFwiZGVwZW5kZW5jaWVzXCI6IHtcbiAgICBcImxpdFwiOiBcIl4yLjAuMFwiLFxuICAgIFwiQGxpdC9yZWFjdGl2ZS1lbGVtZW50XCI6IFwiXjEuMC4wXCIsXG4gICAgXCJsaXQtZWxlbWVudFwiOiBcIl4zLjAuMFwiLFxuICAgIFwibGl0LWh0bWxcIjogXCJeMi4wLjBcIlxuICB9XG59IiwiaGlkZGVuIjp0cnVlfSx7Im5hbWUiOiJ0ZXN0LWNoaWxkLmpzIiwiY29udGVudCI6ImltcG9ydCB7aHRtbCwgTGl0RWxlbWVudH0gZnJvbSAnbGl0JztcblxuY3VzdG9tRWxlbWVudHMuZGVmaW5lKCd0ZXN0LWNoaWxkJywgY2xhc3MgVGVzdENoaWxkIGV4dGVuZHMgTGl0RWxlbWVudCB7XG4gIHN0YXRpYyBwcm9wZXJ0aWVzID0ge1xuICAgIHN0cmluZzogeyB0eXBlOiBTdHJpbmcgfSxcbiAgICBvYmplY3Q6IHsgdHlwZTogT2JqZWN0IH0sXG4gICAgdmVyYm9zZVN0cmluZzogeyB0eXBlOiBTdHJpbmcsIGF0dHJpYnV0ZTogJ3ZlcmJvc2Utc3RyaW5nJyB9LFxuICAgIHZlcmJvc2VPYmplY3Q6IHsgdHlwZTogT2JqZWN0LCBhdHRyaWJ1dGU6ICd2ZXJib3NlLW9iamVjdCcgfSxcbiAgICB2ZXJib3NlT2JqZWN0U2VyaWFsaXplZDogeyB0eXBlOiBTdHJpbmcsIGF0dHJpYnV0ZTogJ3ZlcmJvc2Utb2JqZWN0LXNlcmlhbGl6ZWQnIH0sXG4gIH1cbiAgICAgICAgICAgICAgICAgICAgICBcbiAgY29uc3RydWN0b3IoKSB7XG4gICAgc3VwZXIoKTtcbiAgICB0aGlzLnZlcmJvc2VPYmplY3QgPSB7fTtcbiAgfVxuXG4gIGNvbm5lY3RlZENhbGxiYWNrKCkge1xuICAgIHN1cGVyLmNvbm5lY3RlZENhbGxiYWNrKCk7XG4gICAgdGhpcy52ZXJib3NlT2JqZWN0U2VyaWFsaXplZCA9IEpTT04ucGFyc2UodGhpcy52ZXJib3NlT2JqZWN0U2VyaWFsaXplZCk7XG4gIH1cbiAgICAgIFxuICByZW5kZXIoKSB7XG4gICAgcmV0dXJuIGh0bWxgPGRpdj5cbiAgICAgIDxwPlN0cmluZyBpcyBcIiR7dGhpcy5zdHJpbmd9XCI8L3A-XG4gICAgICA8cD5PYmplY3QgaXMgXCIke0pTT04uc3RyaW5naWZ5KHRoaXMub2JqZWN0KX1cIjwvcD5cbiAgICAgIDxwPlN0cmluZyB3aXRoIGxvbmcgcHJvcCBuYW1lIGlzIFwiJHt0aGlzLnZlcmJvc2VTdHJpbmd9XCI8L3A-XG4gICAgICA8cD5PYmplY3Qgd2l0aCBsb25nIHByb3AgbmFtZSBpcyBcIiR7SlNPTi5zdHJpbmdpZnkodGhpcy52ZXJib3NlT2JqZWN0KX1cIjwvcD5cbiAgICAgIDxwPlNlcmlhbGl6ZWQgb2JqZWN0IHdpdGggbG9uZyBwcm9wIG5hbWUgaXMgXCIke0pTT04uc3RyaW5naWZ5KHRoaXMudmVyYm9zZU9iamVjdFNlcmlhbGl6ZWQpfVwiPC9wPlxuICAgIDwvZGl2PmA7XG4gIH1cbn0pIn1d)

This doesn't actually have to be a property binding anymore though, I'll fix that

Comment (imcdonald-d2l) [reply src/components/editor/rubric-criteria-editor.js:362]: Ah interesting, definitely good to know that. Thanks for explaining and providing the playground example 👍🏻

Comment (devpow112) [reply src/components/editor/rubric-criteria-editor.js:362]: Does setting `.rubricLevelLoaMapping` work. For properties the names have to match I think. The attribute stuff is only used if you don't use `.` notation (at least that's what I thought). Might be better then having to serialize and deserialize stuff on the boundary? Or does that still present the same problem.

Comment (semibran) [reply src/components/editor/rubric-criteria-editor.js]: would make sense, camel case does seem to work. wish this was documented somewhere officially..... this isn't really searchable behavior

Comment (devpow112) [reply src/components/editor/rubric-criteria-editor.js]: Yeah it's kinda annoying that it's not really in the documentation anywhere. I just saw it used in brightspace UI core stuff at some-point. If you want I can add it to our confluence page.

Comment (semibran) [reply src/components/editor/rubric-criteria-editor.js]: I'd appreciate that. Thanks man

**Review by imcdonald-d2l — APPROVED** [review]

**Comment by devpow112** [comment]
> Visual diff tests failed - pull request #1356 has been opened with the updated goldens.

I hate our auto saving 😞. Looks like the diff is just the `Saving...` string.

**Comment by semibran** [comment]
> I hate our auto saving 😞. Looks like the diff is just the `Saving...` string.

Lol tbf the latest commit actually does break stuff. Switching back to property bindings sends it into an infinite request loop since with objects it doesn't check if the property value has changed before reloading the overlay entity and apparently explicit serialization takes care of that automatically. Just been trying to figure out why the borders here aren't updating properly.

image

**Comment by imcdonald-d2l** [comment]
> Just been trying to figure out why the borders here aren't updating properly.

@semibran is this PR ready for review or are there still some loose ends to clean up?

**Comment by semibran** [comment]
> > Just been trying to figure out why the borders here aren't updating properly.
> 
> @semibran is this PR ready for review or are there still some loose ends to clean up?

Should be fixed now 🙂

**Pull request merged**